### PR TITLE
Memcached Implementation

### PR DESCRIPTION
**Pull request by Raj-Shah1**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -57,13 +57,6 @@
       <artifactId>spring-boot-starter-validation</artifactId>
     </dependency>
 
-    <dependency>
```
Comment (AMAN-BARBARIA) [inline pom.xml:60]: Add a version for spring-boot-starter-cache

File: set_env_vars-sample.sh [hunk 2]
```diff
@@ -19,4 +19,8 @@ export DYNAMO_DB_URL='http://dynamodb:8000'
 # Salesforce
 export SALESFORCE_AUTH_URL='https://test.salesforce.com'
 export SALESFORCE_CLIENT_ID=''
-export SALESFORCE_CLIENT_SECRET=''
```
Comment (AMAN-BARBARIA) [inline set_env_vars-sample.sh:22]: We should add these vars for test env vars?

File: src/main/java/com/salessparrow/api/config/CoreConstants.java [hunk 3]
```diff
@@ -34,7 +34,7 @@ public static String awsRegion() {
   }
 
   public static String cacheClusterId() {
-    return System.getenv("CACHE_CLUSTER_ID");
+    return System.getenv("SALESSPARROW_CACHE_CLUSTER_ID");
```
Comment (AMAN-BARBARIA) [inline src/main/java/com/salessparrow/api/config/CoreConstants.java:37]: SALESSPARROW_CACHE_CLUSTER_ID it not defined anwhere?

File: set_env_vars-sample.sh [hunk 4]
```diff
@@ -19,4 +19,8 @@ export DYNAMO_DB_URL='http://dynamodb:8000'
...
\ No newline at end of file
+export SALESFORCE_CLIENT_SECRET=''
+
+# Local Cache
+export LOCAL_CACHE_HOST='localhost'
```
Comment (AMAN-BARBARIA) [inline set_env_vars-sample.sh:25]: Why are we calling it local?

File: src/main/java/com/salessparrow/api/domain/SalesforceOauthToken.java [hunk 5]
```diff
@@ -18,7 +19,7 @@
 @AllArgsConstructor
 @NoArgsConstructor
 @DynamoDBTable(tableName = "salesforce_oauth_tokens")
-public class SalesforceOauthToken {
+public class SalesforceOauthToken implements Serializable {
```
Comment (AMAN-BARBARIA) [inline src/main/java/com/salessparrow/api/domain/SalesforceOauthToken.java:22]: Lets add a comment on why do we need Serializable

File: src/main/java/com/salessparrow/api/config/MemcachedConfig.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  @Override
+  @Bean
+  public CacheManager cacheManager() {
+    SimpleCacheManager cacheManager = new SimpleCacheManager();
+    
+    cacheManager.setCaches(internalCaches());
+    return cacheManager;
+  }
+
+  private Collection<Memcached> internalCaches() {
```
Comment (AMAN-BARBARIA) [inline src/main/java/com/salessparrow/api/config/MemcachedConfig.java:33]: What is the use of internalCaches? 
Do we have to define all the cache keys here?

File: src/main/java/com/salessparrow/api/config/MemcachedConfig.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
+
```
Comment (AMAN-BARBARIA) [inline src/main/java/com/salessparrow/api/config/MemcachedConfig.java:1]: Add comments

File: src/main/java/com/salessparrow/api/utility/Memcached.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+            .setProtocol(ConnectionFactoryBuilder.Protocol.BINARY)
+            .build(),
+          AddrUtil.getAddresses(CoreConstants.memcachedAddress()));
+      } else {
+        BasicAWSCredentials awsCredentials = new BasicAWSCredentials(
+          CoreConstants.awsAccessKeyId(), 
+          CoreConstants.awsSecretAccessKey()
+        );
+
+        AmazonElastiCache amazonElastiCache = AmazonElastiCacheClientBuilder.standard()
```
Comment (AMAN-BARBARIA) [inline src/main/java/com/salessparrow/api/utility/Memcached.java:54]: move to config, similar to how KMS/dynamoDB is defined

File: src/main/java/com/salessparrow/api/utility/Memcached.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
+package com.salessparrow.api.utility;
+
```
Comment (AMAN-BARBARIA) [inline src/main/java/com/salessparrow/api/utility/Memcached.java:2]: Add comments

File: src/main/java/com/salessparrow/api/utility/Memcached.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+import com.amazonaws.services.elasticache.model.CacheCluster;
+import com.amazonaws.services.elasticache.model.DescribeCacheClustersRequest;
+import com.amazonaws.services.elasticache.model.DescribeCacheClustersResult;
+import com.salessparrow.api.config.CoreConstants;
+import com.salessparrow.api.exception.CustomException;
+import com.salessparrow.api.lib.errorLib.ErrorObject;
+
+public class Memcached implements Cache {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(Memcached.class);
```
Comment (AMAN-BARBARIA) [inline src/main/java/com/salessparrow/api/utility/Memcached.java:26]: Should LOGGER variable be lowercase?

File: src/main/java/com/salessparrow/api/lib/globalConstants/CacheConstants.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.springframework.stereotype.Component;
+
+@Component
+public class CacheConstants {
+
+  public static final String SALESFORCE_USER_CACHE = "sf_user";
+  public static final Integer SALESFORCE_USER_CACHE_EXP = 30 * 24 * 60 * 60; // 30 days
+
+  public static final String SALESFORCE_ORG_CACHE = "sf_org";
```
Comment (AMAN-BARBARIA) [inline src/main/java/com/salessparrow/api/lib/globalConstants/CacheConstants.java:11]: Remove it if not used

File: set_env_vars-sample.sh (hunk 4 above)
Comment (Raj-Shah1) [reply set_env_vars-sample.sh:25]: For local only we are using cache host and port, for dev and prod we will be using aws creds

File: set_env_vars-sample.sh (hunk 2 above)
Comment (Raj-Shah1) [reply set_env_vars-sample.sh:22]: Added to test env vars

File: src/main/java/com/salessparrow/api/config/CoreConstants.java (hunk 3 above)
Comment (Raj-Shah1) [reply src/main/java/com/salessparrow/api/config/CoreConstants.java]: Updated the variable name

File: src/main/java/com/salessparrow/api/domain/SalesforceOauthToken.java (hunk 5 above)
Comment (Raj-Shah1) [reply src/main/java/com/salessparrow/api/domain/SalesforceOauthToken.java:28]: Added comment

File: src/main/java/com/salessparrow/api/config/MemcachedConfig.java (hunk 6 above)
Comment (Raj-Shah1) [reply src/main/java/com/salessparrow/api/config/MemcachedConfig.java:33]: We need to define which all cache keys are gonna be used with there expiry here

File: src/main/java/com/salessparrow/api/config/MemcachedConfig.java (hunk 7 above)
Comment (Raj-Shah1) [reply src/main/java/com/salessparrow/api/config/MemcachedConfig.java:1]: Added comments

File: src/main/java/com/salessparrow/api/utility/Memcached.java (hunk 9 above)
Comment (Raj-Shah1) [reply src/main/java/com/salessparrow/api/utility/Memcached.java:2]: Added comments

File: src/main/java/com/salessparrow/api/utility/Memcached.java (hunk 10 above)
Comment (Raj-Shah1) [reply src/main/java/com/salessparrow/api/utility/Memcached.java]: Renamed

File: src/main/java/com/salessparrow/api/lib/globalConstants/CacheConstants.java (hunk 11 above)
Comment (Raj-Shah1) [reply src/main/java/com/salessparrow/api/lib/globalConstants/CacheConstants.java]: Removed unused ones

**Pull request merged**